woocommerce/woocommerce
# API: Add endpoint for fetching location data

Fixes #15080 by adding a new endpoint `/data/locations` returning all available continents, countries, and states; with options to restrict to a continent or country.

```
GET /data # returns an index describing the following endpoints
GET /data/locations # returns all continents, countries and states.
GET /data/locations/<location> # if given a continent code, return all countries and states of that continent.
GET /data/locations/<location> # if given a country code, return all states of a country.
```

This returns data that looks like the following. It uses `WC()->countries->get_*` etc, so any filtered content should be reflected in the output.

```
[
	{
		"code": "NA",
		"name": "North America",
		"countries": [
			{
				"code": "AG",
				"name": "Antigua and Barbuda",
				"states": []
			},
			…
			{
				"code": "US",
				"name": "United States (US)",
				"states": {
					"AL": "Alabama",
					"AK": "Alaska",
					…
```

To test:

- Run the unit test: `phpunit tests/unit-tests/api/data.php`
- Manually test by loading the endpoint, `/data/locations` optionally with continent & country codes, ex `/data/locations/eu`, `/data/locations/us`, or fake data to see an error `/data/locations/aa`
- `/data` lists an "index" view
- All new endpoints should be restricted to store managers
